Managers/UTMVirtualMachine+Drives.m

```diff
@@ -63,42 +70,62 @@ @implementation UTMVirtualMachine (Drives)
 }
 
 - (BOOL)changeMediumForDrive:(UTMDrive *)drive url:(NSURL *)url error:(NSError * _Nullable __autoreleasing *)error {
-    NSData *bookmark = [url bookmarkDataWithOptions:0
+    // assume security scoped, it doesn't hurt even if the file is in the documents directory!
+    BOOL securityScoped = YES;
+    if (securityScoped && ![url startAccessingSecurityScopedResource]) {
+        return NO;
+    }
+    NSUInteger createOptions = securityScoped ? ( 1 << 11 ) : 0;
+    NSData *bookmark = [url bookmarkDataWithOptions:createOptions
                      includingResourceValuesForKeys:nil
                                       relativeToURL:nil
                                               error:error];
     if (!bookmark) {
         if (error) {
-            *error = [NSError errorWithDomain:kUTMErrorDomain code:-1 userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"Failed create bookmark.", "UTMVirtualMachine+Drives")}];
+            *error = [NSError errorWithDomain:kUTMErrorDomain code:-1 userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"Failed to create bookmark.", "UTMVirtualMachine+Drives")}];
         }
         return NO;
     }
-    if (![self checkSandboxAccess:url error:error]) {
+    // save configuration
+    [self.configuration setImagePath:url.path forIndex:drive.index];
+    bool saved = [self saveUTMWithError:error];
+    if (!saved) {
         return NO;
     }
-    [self.viewState setBookmark:bookmark path:url.path forRemovableDrive:drive.name persistent:NO];
     if (!self.qemu.isConnected) {
-        return YES; // not ready yet
+        [self.viewState setBookmark:bookmark path:url.path forRemovableDrive:drive.name];
+        return YES; // VM not running
     } else {
-        return [self changeMediumForDriveInternal:drive bookmark:bookmark persistent:NO error:error];
+        bool ret = [self changeMediumForDriveInternal:drive bookmark:bookmark securityScoped:securityScoped error:error];
+        if (ret) {
+            [self.viewState setBookmark:bookmark path:url.path forRemovableDrive:drive.name];
+        }
+        return ret;
     }
 }
 
-- (BOOL)changeMediumForDriveInternal:(UTMDrive *)drive bookmark:(NSData *)bookmark persistent:(BOOL)persistent error:(NSError * _Nullable __autoreleasing *)error {
+- (BOOL)changeMediumForDriveInternal:(UTMDrive *)drive bookmark:(NSData *)bookmark securityScoped:(BOOL)securityScoped error:(NSError * _Nullable __autoreleasing *)error {
     __block BOOL ret = NO;
     dispatch_semaphore_t sema = dispatch_semaphore_create(0);
     [self.system accessDataWithBookmark:bookmark
-                         securityScoped:persistent
+                         securityScoped:securityScoped
                              completion:^(BOOL success, NSData *newBookmark, NSString *path) {
         if (success) {
-            [self.viewState setBookmark:newBookmark path:path forRemovableDrive:drive.name persistent:YES];
```

Comment:
This is actually a workaround for a nasty bug on macOS sandbox. Basically, when you create a bookmark on the main process it cannot be used on the XPC process. When the user first selects the file, you get a “lite” bookmark that you can pass to the XPC process and it can handle it fine. However, if you turn that “lite” bookmark into a full bookmark and pass it to the XPC process, it won’t work (by design). The only way around this is to first pass the “lite” bookmark to the XPC process, then the XPC process creates the full bookmark, and passes it BACK to the main app which saves it to the plist.

Reply:
This is why I have this “persistent” flag passed around. The “lite” bookmarks are persistent=NO.

Reply:
In case it wasn’t clear, “lite” bookmarks are not persistent.

Reply:
Oh that is what `persistent` means. It wasn't clear to me from the code.